arch/arm64: Use large code model when text or data base is above 4GB boundary

keith-packard opened the pull request:
The TLS initialization code has some clever use of linker-generated values to avoid doing computation at runtime. That works great on most targets, but on arm64, the linker can end up unable to generate working code in some memory configurations.
    
Avoid this by using the large code model when the memory configuration places text or data above the 4GB boundary.

keith-packard committed:
tests/arm64: Run high_addr tests using picolibc

Picolibc enables TLS, which causes relocation errors with some of these
tests without changing the compiler code model.

Signed-off-by: Keith Packard <keithp@keithp.com>

andyross reviewed (commented):
Is this a linker bug?  What's the root cause here?

A quick grep shows that these aren't relocations in the generated code, they're just PROVIDE'd based on SIZEOF/ALIGNOF of the already linked section.  If whatever linker is in use isn't working here, there's probably a deeper bug at work.  For sure there's nothing arch-specific related to arm64 that I can see.

Has TLS never worked on arm64 before?  What are the existing tests doing that works, or do we have a hole in coverage?

povergoing commented:
Correct me if I am wrong, arm64 should not be special? Sometimes, in terms of the alignment issue, using `SECTION_DATA_PROLOGUE(data,, ALIGNMENT)` may make things better. Or maybe it is caused by a high memory address which is a bit weird?

keith-packard committed:
arch/arm64: Enable 'large' code model for large targets

Targets with text or data addresses above the 4GB boundary may need to use
the large code model to ensure relocations in the linker work correctly.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> Is this a linker bug? What's the root cause here?

Ok, so I went and looked at the assembly output for this case and that helped explain the issue.  Arm64 (among other architectures) has a very limited range of values that can be used as addresses when being relocated by the linker. In particular, when you use __tdata_size, you get:
```
	adrp	x2, __tdata_size
	add	x2, x2, :lo12:__tdata_size
```
`adrp` takes an address *relative to the PC* in the range +/-4GB and loads the upper 52 bits into the register. The `add` instruction adds the low 12 bits of the same address to the register. This avoids the use of code space for address constants as is done on 32-bit ARM targets. Risc-V has the same design, with an even more limited range of values.

The cases which fail are all of the `high_addr` tests, which place the text starting at 0x2_0088_0000.

Recalling this same issue occurs on Risc-V, I remembered that there were multiple 'code models' available for that architecture and went and read the GCC man page:
```       -mcmodel=tiny
           Generate code for the tiny code model.  The program and its
           statically defined symbols must be within 1MB of each other.
           Programs can be statically or dynamically linked.

       -mcmodel=small
           Generate code for the small code model.  The program and its
           statically defined symbols must be within 4GB of each other.
           Programs can be statically or dynamically linked.  This is the
           default code model.

       -mcmodel=large
           Generate code for the large code model.  This makes no assumptions
           about addresses and sizes of sections.  Programs can be statically
           linked only.  The -mcmodel=large option is incompatible with
           -mabi=ilp32, -fpic and -fPIC.
```
To make these examples work, we just need to use the 'large' code model. I'm updating the patch to use that instead of hacking the code.

> A quick grep shows that these aren't relocations in the generated code, they're just PROVIDE'd based on SIZEOF/ALIGNOF of the already linked section. If whatever linker is in use isn't working here, there's probably a deeper bug at work. For sure there's nothing arch-specific related to arm64 that I can see.

Yeah, it's related to how the compiler generates code for relocatable values.

> Has TLS never worked on arm64 before? What are the existing tests doing that works, or do we have a hole in coverage?

It only fails on the high_addr tests, and only when TLS is enabled because that's when we get this large spread of relocation values. I'm a bit surprised it worked otherwise; I would have expected there to be data addresses relocated in the code that would fail here.

keith-packard commented:
> Correct me if I am wrong, arm64 should not be special? Sometimes, in terms of the alignment issue, using `SECTION_DATA_PROLOGUE(data,, ALIGNMENT)` may make things better. Or maybe it is caused by a high memory address which is a bit weird?

Good guess -- the 'high_addr' tests were the ones failing because they were being built with the small code model. Switching those to the large model fixes the problem.

carlocaione commented on arch/arm64/core/CMakeLists.txt:
```diff
@@ -15,6 +15,18 @@ zephyr_library_sources(
+string(LENGTH "x${CONFIG_SRAM_BASE_ADDRESS}" SRAM_LENGTH)
+string(LENGTH "x${CONFIG_KERNEL_VM_BASE}" KERNEL_VM_LENGTH)
+
+if(${SRAM_LENGTH} GREATER 11 OR ${KERNEL_VM_LENGTH} GREATER 11)
+  zephyr_cc_option(-mcmodel=large)
+endif()
```
Is there any issue to have this always set for ARM64?

povergoing replied:
I found this:
```
       -mcmodel=tiny
           Generate code for the tiny code model.  The program and its statically defined symbols must be within 1MB of each other.  Programs can be statically or dynamically linked.

       -mcmodel=small
           Generate code for the small code model.  The program and its statically defined symbols must be within 4GB of each other.  Programs can be statically or dynamically linked.  This is the default code model.

       -mcmodel=large
           Generate code for the large code model.  This makes no assumptions about addresses and sizes of sections.  Programs can be statically linked only.
```

povergoing replied:
AArch64 Options in GCC

carlocaione replied:
Right, but Zephyr is by default only statically linked so why not set it all the times?

povergoing replied:
Yes, I think no problem

andyross replied:
Presumably there's a code size impact; the relocations become physically larger and presumably require different instruction formats.

keith-packard replied:
Yes, the generated code is larger and slower. But, it's also going to work, while the other models may have distant relocations which the linker will fail to resolve. What we want is a way to tell the compiler that *code* references are near and *data* references may be far. Not sure that's one of the options.

carlocaione replied:
alright I propose to go with this for now.

andyross reviewed (approved):
Nice, thanks.  This makes much more sense to me.

I've been looking at linkage and relocation recently (treating my current bout of unrestrained hubris) and it's really surprising the extent to which arm64 made a big mess of its relocation strategy.  They had every opportunity to design it cleanly from scratch, and ended up with what seems to be a taped up forward port of a 32 bit architecture?  Ick.

+1 for the explanation.  I'll leave it to arm64 experts to determine if there isn't a better way than manual management of address spaces.

npitre commented:
FWIW this is likely the same issue as #32053

keith-packard commented:
> FWIW this is likely the same issue as #32053

Yup. Seems like we can either use -mcmodel=large or remove all link-time computed sizes/alignments. The sizes are easy, I'm not entirely sure how to deal with alignments though.

andyross commented:
> The sizes are easy, I'm not entirely sure how to deal with alignments though.

Oh, that's easy.  All you need to do is create a new pointer-sized integer in the linker plugin containing the value you just computed based on the first section, then insert it into .rodata as a named symbol you can reference from code.

What, your linker can't do that?  Do I have the slide deck for you...  (no code yet though)
